File: Solutions/921. Minimum Add to Make Parentheses Valid/921. Minimum Add to Make Parentheses Valid.java
/**
 * 
 */
package github.com.ErnestL1n;

/**
 * @author https://github.com/ErnestL1n
 *
 */
//Leetcode.921
public class Minimum_Add_to_Make_Parentheses_Valid {

	/**
	 * @param args
	 */
	public static int minAddToMakeValid(String S) {
		int answer=0,count=0;
		for(int i=0;i<S.length();i++) {
			if(S.charAt(i)=='(')++count;
			if(S.charAt(i)==')')--count;
			//can't not repair mistakes,just add into answer and restart
			if(S.charAt(i)==')'&&count<=0) {
				answer+=Math.abs(count);
				count=0;
			}
		}
		answer+=count;
		return answer;
	}
	public static void main(String[] args) {
		String S="()))((";
		System.out.println("Given a string S of '(' and ')' parentheses : "+S);
		System.out.println("Minimum number of parentheses we must add to make the resulting string valid is "+minAddToMakeValid(S));

	}
	

}

File: Solutions/921. Minimum Add to Make Parentheses Valid/921. Minimum Add to Make Parentheses Valid.cpp
class Solution {
public:
    int minAddToMakeValid(string s) {
        int res=0,cnt=0;
        for(int i=0;i<s.size();++i){
            if(s[i]=='(')
                ++cnt;
            if(s[i]==')')
                --cnt;
            if(s[i]==')'&&cnt<0){
                res+=abs(cnt);
                cnt=0;
            }
        }
        res+=cnt;
        return res;
    }
};